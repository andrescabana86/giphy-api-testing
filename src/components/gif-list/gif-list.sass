.gif-list
  display: grid
  grid-template-columns: repeat(auto-fit, min(96px, 128px))
  grid-auto-columns: 1fr
  grid-auto-rows: 1fr
  grid-gap: 5px
  width: 600px
  min-width: 320px
  margin: 0 auto
  justify-content: center