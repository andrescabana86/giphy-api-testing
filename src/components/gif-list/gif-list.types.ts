import { TResult, TResultTrending } from '@api/types/giphy.types'

export type TGifListProps = {
  list: (TResult | TResultTrending)[]
  loading: boolean
}
