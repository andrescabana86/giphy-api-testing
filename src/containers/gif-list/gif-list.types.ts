import { TResult, TResultTrending } from '@api/types/giphy.types'

export type TGifListContainerProps<T> = {
  children(container: T): JSX.Element
}

export type TGifListContainerState = {
  favorites: (TResult | TResultTrending)[]
  list: (TResult | TResultTrending)[]
  loading: boolean
}
