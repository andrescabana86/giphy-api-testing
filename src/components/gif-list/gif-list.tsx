import React, { useContext } from 'react'
import { GifPreview } from '@components/gif-preview/gif-preview'
import { FavoritesGifsContext } from '@containers/gif-list/gif-list'
import { TGifListProps } from './gif-list.types'
import './gif-list.sass'

export const GifList = React.memo((props: TGifListProps) => {
  const { list } = props
  const FavoritesContext = useContext(FavoritesGifsContext)
  const { favorites } = FavoritesContext
  return (
    <section className="gif-list" data-testid="gif-list">
      {list.map((gif) => (
        <GifPreview
          id={gif.id}
          key={gif.id}
          selected={!!favorites.find((favorite) => favorite.id === gif.id)}
          url={gif.url}
        />
      ))}
    </section>
  )
})
