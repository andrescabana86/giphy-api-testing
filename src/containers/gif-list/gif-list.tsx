import React, { createContext, PureComponent } from 'react'
import { getTrending } from '@api/trending/trending'
import { searchGiphy } from '@api/search/search'
import { TResult, TResultTrending } from '@api/types/giphy.types'
import { TGifListContainerProps, TGifListContainerState } from './gif-list.types'

export const FavoritesGifsContext = createContext<{
  favorites: (TResult | TResultTrending)[]
  saveFavorite?(id: string): void
  removeFavorite?(id: string): void
}>({ favorites: [] })

export class GifListContainer extends PureComponent<
  TGifListContainerProps<GifListContainer>,
  TGifListContainerState
> {
  state: TGifListContainerState = {
    favorites: [],
    list: [],
    loading: true,
  }

  componentDidMount() {
    this.getTrendingList()
  }

  onSearch = (query: string) => {
    this.setState({ loading: true })
    searchGiphy(query).then((results) => {
      this.setState({ loading: false, list: results })
    })
  }

  getTrendingList = () => {
    this.setState({ loading: true })
    getTrending().then((results) => {
      this.setState({ list: results, loading: false })
    })
  }

  onClear = () => {
    this.getTrendingList()
  }

  saveFavorite = (id: string) => {
    const { list, favorites } = this.state
    const favorite = list.find((item) => item.id === id)
    if (favorite) {
      this.setState({
        favorites: [...favorites, favorite],
      })
    }
  }

  removeFavorite = (id: string) => {
    const { favorites } = this.state
    const updatedList = favorites.filter((item) => item.id !== id)
    this.setState({
      favorites: [...updatedList],
    })
  }

  render() {
    const { favorites } = this.state

    return (
      <FavoritesGifsContext.Provider
        value={{ favorites, saveFavorite: this.saveFavorite, removeFavorite: this.removeFavorite }}
      >
        {this.props.children(this)}
      </FavoritesGifsContext.Provider>
    )
  }
}
