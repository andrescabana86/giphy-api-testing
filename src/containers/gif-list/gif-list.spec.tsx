import React from 'react'
import { disableFetchMocks, enableFetchMocks } from 'jest-fetch-mock'
import { render, cleanup } from '@testing-library/react'
import { GifListContainer } from './gif-list'

beforeAll(enableFetchMocks)
afterAll(disableFetchMocks)

describe('GifListContainer', () => {
  afterEach(cleanup)

  it('should render GifListContainer component', () => {
    // Arrange
    fetchMock.mockResponseOnce(JSON.stringify({ data: [] }))

    // Act
    const { asFragment } = render(
      <GifListContainer>{() => <span>this is fake children</span>}</GifListContainer>,
    )

    // Assert
    expect(asFragment()).toMatchSnapshot()
  })
})
