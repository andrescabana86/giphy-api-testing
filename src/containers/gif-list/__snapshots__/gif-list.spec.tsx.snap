// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifListContainer should render GifListContainer component 1`] = `
<DocumentFragment>
  <span>
    this is fake children
  </span>
</DocumentFragment>
`;
